File: Tests/CeyradTests/SharedVectorTests.swift
import XCTest

@testable import Ceyrad

/// `spec/activity_vectors.json` を読んで、macOS版の組み立てが仕様どおりか確かめる。
///
/// 同じファイルを `windows/tests/shared_vector_tests.rs` も読む。片方だけを直すと
/// もう片方が落ちる——それがこのテストの唯一の目的で、Activityの中身そのものは
/// `ActivityBuilderTests` が個別に見ている。
final class SharedVectorTests: XCTestCase {
    private struct Spec: Decodable {
        var cases: [Case]
    }

    private struct Case: Decodable {
        var name: String
        var track: TrackSpec
        var playerState: String
        var catalog: CatalogSpec?
        var settings: SettingsSpec
        var nowUnix: Double
        var expected: [String: JSONValue]
    }

    private struct TrackSpec: Decodable {
        var name: String
        var artist: String
        var album: String
        var durationSec: Double?
        var positionSec: Double?
        var positionSampledAtUnix: Double
    }

    private struct CatalogSpec: Decodable {
        var songURL: String?
        var artistURL: String?
        var albumURL: String?
        var artworkURL: String?
    }

    private struct SettingsSpec: Decodable {
        var button1Type: String?
        var button2Type: String?
        var button1Label: String?
        var button2Label: String?
        var customURL: String?
        var repositoryURL: String?
        var badgeLabel: String?
        var pauseHideMinutes: Int?
    }

    private static let suiteName = "CeyradTests.SharedVectors"

    /// リポジトリ内の `spec/` は、テストバンドルではなくソースからの相対で引く。
    /// SwiftPMのリソース宣言を足すとビルド設定がもう1つ増えるだけで、
    /// このファイルは実行時ではなくテスト時にしか要らない。
    private static var specURL: URL {
        URL(fileURLWithPath: #filePath)  // Tests/CeyradTests/SharedVectorTests.swift
            .deletingLastPathComponent()  // Tests/CeyradTests
            .deletingLastPathComponent()  // Tests
            .deletingLastPathComponent()  // <repo>
            .appendingPathComponent("spec/activity_vectors.json")
    }

    func testEveryVectorBuildsExactlyWhatTheSpecSays() throws {
        let data = try Data(contentsOf: Self.specURL)
        let spec = try JSONDecoder().decode(Spec.self, from: data)
        XCTAssertFalse(spec.cases.isEmpty, "ベクタが空: 読み込みが壊れている")

        let defaults = UserDefaults(suiteName: Self.suiteName)!
        defer { defaults.removePersistentDomain(forName: Self.suiteName) }

        for testCase in spec.cases {
            defaults.removePersistentDomain(forName: Self.suiteName)
            let settings = SettingsStore(defaults: defaults)
            apply(testCase.settings, to: settings)

            let activity = ActivityBuilder.build(
                track: track(from: testCase.track),
                playerState: playerState(testCase.playerState),
                catalog: testCase.catalog.map(catalog(from:)),
                settings: settings,
                now: Date(timeIntervalSince1970: testCase.nowUnix)
            )

            let expected = testCase.expected.mapValues(\.any)
            XCTAssertTrue(
                ActivityBuilder.isEquivalent(activity, expected)
                    && (activity as NSDictionary).isEqual(to: expected),
                """
                \(testCase.name)
                期待: \(sorted(expected))
                実際: \(sorted(activity))
                """
            )
        }
    }

    // MARK: - 組み立て

    private func track(from spec: TrackSpec) -> TrackInfo {
        var track = TrackInfo(
            name: spec.name, artist: spec.artist, album: spec.album,
            durationSec: spec.durationSec, positionSec: spec.positionSec
        )
        track.positionSampledAt = Date(timeIntervalSince1970: spec.positionSampledAtUnix)
        return track
    }

    private func catalog(from spec: CatalogSpec) -> CatalogInfo {
        CatalogInfo(
            songURL: spec.songURL, artistURL: spec.artistURL,
            albumURL: spec.albumURL, artworkURL: spec.artworkURL
        )
    }

    private func playerState(_ raw: String) -> PlayerState {
        switch raw {
        case "playing": return .playing
        case "paused": return .paused
        default: return .stopped
        }
    }

    private func apply(_ spec: SettingsSpec, to settings: SettingsStore) {
        applyButtons(spec, to: settings)
        if let url = spec.customURL { settings.customURL = url }
        if let url = spec.repositoryURL { settings.repositoryURL = url }
        if let raw = spec.badgeLabel { settings.badgeLabel = badgeLabel(raw) }
        if let minutes = spec.pauseHideMinutes { settings.pauseHideMinutes = minutes }
    }

    private func applyButtons(_ spec: SettingsSpec, to settings: SettingsStore) {
        let types = [(ButtonSlot.one, spec.button1Type), (.two, spec.button2Type)]
        for (slot, raw) in types {
            guard let raw else { continue }
            guard let type = LinkType(rawValue: raw) else {
                return XCTFail("未知のリンク先: \(raw)")
            }
            settings.setButtonType(slot, type)
        }
        let labels = [(ButtonSlot.one, spec.button1Label), (.two, spec.button2Label)]
        for (slot, label) in labels {
            if let label { settings.setButtonLabel(slot, label) }
        }
    }

    private func badgeLabel(_ raw: String) -> BadgeLabelType {
        switch raw {
        case "appName": return .appName
        case "artist": return .artist
        case "track": return .track
        default:
            XCTFail("未知のバッジ指定: \(raw)")
            return .artist
        }
    }

    private func sorted(_ activity: [String: Any]) -> String {
        activity.keys.sorted().map { "\($0)=\(activity[$0]!)" }.joined(separator: " ")
    }
}

/// JSONをそのまま`[String: Any]`に落とすための最小のデコーダ。
/// `JSONSerialization`だとNSNumberの取り回しが型ごとに変わるため、
/// Int/Double/String/配列/辞書を明示的に受ける。
private enum JSONValue: Decodable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
    case array([JSONValue])
    case object([String: JSONValue])
    case null

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Int.self) {
            self = .int(value)
        } else if let value = try? container.decode(Double.self) {
            self = .double(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([JSONValue].self) {
            self = .array(value)
        } else {
            self = .object(try container.decode([String: JSONValue].self))
        }
    }

    var any: Any {
        switch self {
        case .string(let value): return value
        case .int(let value): return value
        case .double(let value): return value
        case .bool(let value): return value
        case .array(let values): return values.map(\.any)
        case .object(let values): return values.mapValues(\.any)
        case .null: return NSNull()
        }
    }
}

File: windows/tests/shared_vector_tests.rs
//! Reads `spec/activity_vectors.json` and checks the Windows build against it.
//!
//! The macOS build reads the same file from
//! `Tests/CeyradTests/SharedVectorTests.swift`. Changing one implementation and
//! not the other makes this fail — which is the only purpose of the file. What
//! each field means is covered case by case in `activity_builder_tests.rs`.

use std::path::PathBuf;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use serde::Deserialize;
use serde_json::{Map, Value};

use ceyrad::core::activity_builder::build;
use ceyrad::core::models::{CatalogInfo, PlayerState, TrackInfo};
use ceyrad::core::settings_model::{BadgeLabelType, ButtonSlot, LinkType, Settings};

#[derive(Deserialize)]
struct Spec {
    cases: Vec<Case>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct Case {
    name: String,
    track: TrackSpec,
    player_state: String,
    catalog: Option<CatalogSpec>,
    #[serde(default)]
    settings: SettingsSpec,
    now_unix: f64,
    expected: Map<String, Value>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct TrackSpec {
    name: String,
    artist: String,
    album: String,
    duration_sec: Option<f64>,
    position_sec: Option<f64>,
    position_sampled_at_unix: f64,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct CatalogSpec {
    #[serde(default, rename = "songURL")]
    song_url: Option<String>,
    #[serde(default, rename = "artistURL")]
    artist_url: Option<String>,
    #[serde(default, rename = "albumURL")]
    album_url: Option<String>,
    #[serde(default, rename = "artworkURL")]
    artwork_url: Option<String>,
}

#[derive(Default, Deserialize)]
#[serde(rename_all = "camelCase")]
struct SettingsSpec {
    #[serde(default)]
    button1_type: Option<String>,
    #[serde(default)]
    button2_type: Option<String>,
    #[serde(default)]
    button1_label: Option<String>,
    #[serde(default)]
    button2_label: Option<String>,
    #[serde(default, rename = "customURL")]
    custom_url: Option<String>,
    #[serde(default, rename = "repositoryURL")]
    repository_url: Option<String>,
    #[serde(default)]
    badge_label: Option<String>,
    #[serde(default)]
    pause_hide_minutes: Option<i32>,
}

/// The spec lives beside the two implementations, not inside either of them.
fn spec_path() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("windows/ has a parent")
        .join("spec/activity_vectors.json")
}

#[test]
fn every_vector_builds_exactly_what_the_spec_says() {
    let raw = std::fs::read_to_string(spec_path()).expect("read spec/activity_vectors.json");
    let spec: Spec = serde_json::from_str(&raw).expect("parse spec/activity_vectors.json");
    assert!(!spec.cases.is_empty(), "no vectors: the read is broken");

    for case in &spec.cases {
        let activity = build(
            &track(&case.track),
            player_state(&case.player_state),
            case.catalog.as_ref().map(catalog).as_ref(),
            &settings(&case.settings),
            unix(case.now_unix),
        );
        assert_eq!(
            Value::Object(activity),
            Value::Object(case.expected.clone()),
            "{}",
            case.name
        );
    }
}

fn unix(seconds: f64) -> SystemTime {
    UNIX_EPOCH + Duration::from_secs_f64(seconds)
}

fn track(spec: &TrackSpec) -> TrackInfo {
    let mut track = TrackInfo::new(&spec.name, &spec.artist, &spec.album);
    track.duration_sec = spec.duration_sec;
    track.position_sec = spec.position_sec;
    track.position_sampled_at = unix(spec.position_sampled_at_unix);
    track
}

fn catalog(spec: &CatalogSpec) -> CatalogInfo {
    CatalogInfo {
        song_url: spec.song_url.clone(),
        artist_url: spec.artist_url.clone(),
        album_url: spec.album_url.clone(),
        artwork_url: spec.artwork_url.clone(),
    }
}

fn player_state(raw: &str) -> PlayerState {
    match raw {
        "playing" => PlayerState::Playing,
        "paused" => PlayerState::Paused,
        _ => PlayerState::Stopped,
    }
}

fn settings(spec: &SettingsSpec) -> Settings {
    let mut settings = Settings::default();
    if let Some(raw) = &spec.button1_type {
        settings.set_button_type(ButtonSlot::One, link_type(raw));
    }
    if let Some(raw) = &spec.button2_type {
        settings.set_button_type(ButtonSlot::Two, link_type(raw));
    }
    if let Some(label) = &spec.button1_label {
        settings.set_button_label(ButtonSlot::One, label);
    }
    if let Some(label) = &spec.button2_label {
        settings.set_button_label(ButtonSlot::Two, label);
    }
    if let Some(url) = &spec.custom_url {
        settings.custom_url = url.clone();
    }
    if let Some(url) = &spec.repository_url {
        settings.repository_url = url.clone();
    }
    if let Some(raw) = &spec.badge_label {
        settings.badge_label = match raw.as_str() {
            "appName" => BadgeLabelType::AppName,
            "artist" => BadgeLabelType::Artist,
            "track" => BadgeLabelType::Track,
            other => panic!("unknown badge label: {other}"),
        };
    }
    if let Some(minutes) = spec.pause_hide_minutes {
        settings.pause_hide_minutes = minutes;
    }
    settings
}

fn link_type(raw: &str) -> LinkType {
    match raw {
        "song" => LinkType::Song,
        "artist" => LinkType::Artist,
        "album" => LinkType::Album,
        "custom" => LinkType::Custom,
        "repository" => LinkType::Repository,
        "disabled" => LinkType::Disabled,
        other => panic!("unknown link type: {other}"),
    }
}
